### DISCUSSION TITLE
How to sync downloaded data efficiently?

### DISCUSSION THREAD

**Discussion by jasonyslai** [Azure/azure-mobile-apps#240] [Q&A] [answered]
Hi,

`await table.PullAsync(queryName, table.Where(x => installedIpList.Contains(x.IpID))); // AzureCloudTable.cs` 

Since where condition statement has a char limit restriction, and have to split PullAsync as below:

```
async Task syncOfflineCahceAsyncInstalled<T>(List<List<Guid>> eachGuidList, int start, int step) where T : Abstractions.TableData
{
	var eachStep = step / (double)eachGuidList.Count;
	await Task.Run(() => { //Split list into sub list and sync
		foreach (var eachGuidList in eachGuidList)
		{
			cloudService.SyncOfflineCacheAsyncInstalled<T>(
				eachGuidList
			);
			UpdateProgress += eachStep;
			Task.Delay(10);
		}
	});
}
```
When the Guid List grows, it takes too long to update. (Guid can be existed more once (1:N) on the same table)
Is that a better way to accomplish this requirement?
regards,
Jason

**Comment by adrianhall** [comment]
Can you provide a fuller use case.  The data model and usage, and what you actually want to sync across?

It's likely that a re-work of the data model with synchronization in mind is needed.  However, it's impossible to say without more information about your application.

**Comment by jasonyslai** [comment, marked as answer]
Hi @adrianhall ,

Data Model:
  Subscriptions which has different service items.
  Each User subscribes his/her favorite subscriptions. 

Usage:
  When user press sync button, It have to sync subscriptions which user subscribed. (service items need to be synced as well.)

**Reply by adrianhall** [reply]
Please actually copy the data models in so I can take a look at the specifics.

**Reply by jasonyslai** [reply]
Hi @adrianhall,
  FYI, Thanks!
```
public class InsurancePolicies : EntityData
{
	//If true, data would be synced.
	public bool IsSubscribed { get; set; }

	[Index("IX_InsurancePolicies_GUID", 1, IsUnique = true)]
	/// <summary>
	/// pID
	/// </summary>
	public Guid Guid { get; set; }

	[MaxLength(50)]
	public string Name { get; set; } = "";
}
```

```
public class BasicCoeffLists : EntityData
{
	/// <summary>
	/// pID
	/// </summary>
	[Index("IX_BasicCoeffList", 1, IsUnique = true)]
	public Guid IpID { get; set; }
	
	//public decimal Dir { get; set; }
	[Index("IX_BasicCoeffList", 2, IsUnique = true)]
	public short Annual { get; set; }
	[Index("IX_BasicCoeffList", 3, IsUnique = true)]
	public short AgeFrom { get; set; }
	[Index("IX_BasicCoeffList", 4, IsUnique = true)]
	public Sex Sex { get; set; }

	public string CashValue { get; set; }
	public string PayOff { get; set; }
	public string GernalGuarantee { get; set; }
	public string LoanAmount { get; set; }
	public string Annuity { get; set; }
	public string AccidentGurantee { get; set; }
	public string SpecificAccidentGurantee { get; set; }
}
```

**Reply by adrianhall** [reply]
So, firstly, this model definition won't work for multi-user situations, and you don't have appropriate fields for doing incremental sync, which means all data is pulled all the time.  There is also a difficulty in getting a select statement that covers all subscriptions.

Funnily enough, doing this in ASP.NET Core land is relatively easy (and impossible in .NET Framework service).  That's because we've made it easier to write a repository to use.  Look at the code for Microsoft.AspNetCore.Datasync.EFCore - you basically define a new entity table data class (which is an abstract class for Id/UpdatedAt/Version/etc.) and then a set of CRUDL methods to return the data you want.  You can write the CRUDL methods to do the JOIN behind the scenes, returning the relevant data to you.

Once that is done, you can then have a regular EFCore table that gives the insurance policies the user is subscribed to (as a personal table), and the basiccoefflists table that is the join.  Both tables are "personal" (they both have userId fields that are not exposed) - it's just one of them is a join and the other is a regular table.

In either case, you will want to set up incremental sync in order to cut down on the amount of data transferred.

And, yes, I'm writing a series of blogs to support the launch of the ASP.NET Core edition.  I'll include this use case in the sequence to show how it's done.

**Reply by jasonyslai** [reply]
Thanks, @adrianhall.
It would be great to have use case for study.
Where can I find this repo? Microsoft.AspNetCore.Datasync
and use case will be uploaded here? https://azure.github.io/azure-mobile-apps/howto/server/aspnetcore/

**Reply by adrianhall** [reply]
The Microsoft.AspNetCore.Datasync isn't released yet (will be next week).  The repo (if you want to build them yourself) is this one - https://github.com/azure/azure-mobile-apps - just look in the sdk/dotnet folder for the SDK code, and take a look at the samples/TodoApp for a sample .NET 6 service backend.

I really need to do an introductory series of posts for the docs though.  So, stay tuned for that.

**Reply by adrianhall** [reply]
I've done the introductory series.  You can find the one on the repository here: https://adrianhall.github.io/azure/2021/11/12/zumo-irepository/

**Reply by jasonyslai** [reply]
Thanks @adrianhall ! I am going to study it and refactor existing code =)!
